### ISSUE TITLE
[data] [streaming] Locality support for actor pool compute strategy

### ISSUE THREAD

**Issue by ericl** [ray-project/ray#31909]
Currently, actor pool operators do not benefit from Ray's locality scheduling. This is because locality scheduling only applies for tasks. For multi-step inference pipelines, this can enormously improve throughput, since perfect locality scheduling would mean almost no data movement in the cluster, compared to 100% data movement.

We could implement locality scheduling specifically for ActorPoolMapOperator, by preferentially dispatching tasks to actors that are already co-located with the ready output bundle.

There are a few complexities on the implementation side:
- Because bundles may be queued outside the ActorPoolMapOperator, we need to allow the actor pool to preferentially pick an input bundle to take as input, rather than always popping the leftmost inqueue in `dispatch_next_task()`. The actor pool map operator would pick an input that matches the locality of an actor with currently free execution slots.
- We need to efficiently query the locations of objects in the actor pool inqueue (batch query them), since `get_object_locations()` incurs RPCs. We could probably batch query all unknown bundle locations when we need to decide which bundle to dispatch.

In short, this change may require minor API extensions to the operator/executor interface to enable locality-aware bundle selection.
